common/math/math.go

```diff
@@ -238,3 +244,233 @@ func SharpeRatio(movementPerCandle []float64, riskFreeRatePerInterval, average f
+	fX, _ := x.Float64()
+	fY, _ := y.Float64()
+	pow := math.Pow(fX, fY)
+	return decimal.NewFromFloat(pow)
+}
+
+// DecimalFinancialGeometricMean is a modified geometric average to assess
+// the negative returns of investments. It accepts It adds +1 to each
+// This does impact the final figures as it is modifying values
+// It is still ultimately calculating a geometric average
+// which should only be compared to other financial geometric averages
+func DecimalFinancialGeometricMean(values []decimal.Decimal) (decimal.Decimal, error) {
+	if len(values) == 0 {
+		return decimal.Zero, errZeroValue
+	}
+	product := 1.0
+	for i := range values {
+		if values[i].LessThan(decimal.NewFromInt(-1)) {
+			// cannot lose more than 100%, figures are incorrect
+			// losing exactly 100% will return a 0 value, but is not an error
+			return decimal.Zero, errNegativeValueOutOfRange
+		}
+		// as we cannot have negative or zero value geometric numbers
+		// adding a 1 to the percentage movements allows for differentiation between
+		// negative numbers (eg -0.1 translates to 0.9) and positive numbers (eg 0.1 becomes 1.1)
+		modVal, _ := values[i].Add(decimal.NewFromInt(1)).Float64()
+		product *= modVal
+	}
+	prod := 1 / float64(len(values))
+	geometricPower := math.Pow(product, prod)
+	if geometricPower > 0 {
+		// we minus 1 because we manipulated the values to be non-zero/negative
+		geometricPower--
+	}
+	return decimal.NewFromFloat(geometricPower), nil
+}
+
+// DecimalArithmeticMean is the basic form of calculating an average.
+// Divide the sum of all values by the length of values
+func DecimalArithmeticMean(values []decimal.Decimal) (decimal.Decimal, error) {
+	if len(values) == 0 {
+		return decimal.Zero, errZeroValue
+	}
+	var sumOfValues decimal.Decimal
+	for x := range values {
+		sumOfValues = sumOfValues.Add(values[x])
+	}
+	return sumOfValues.Div(decimal.NewFromInt(int64(len(values)))), nil
+}
+
+// DecimalSortinoRatio returns sortino ratio of backtest compared to risk-free
+func DecimalSortinoRatio(movementPerCandle []decimal.Decimal, riskFreeRatePerInterval, average decimal.Decimal) (decimal.Decimal, error) {
+	totalIntervals := decimal.NewFromInt(int64(len(movementPerCandle)))
+	if totalIntervals.IsZero() {
+		return decimal.Zero, errZeroValue
+	}
+	totalNegativeResultsSquared := decimal.Zero
+	for x := range movementPerCandle {
+		if movementPerCandle[x].Sub(riskFreeRatePerInterval).LessThan(decimal.Zero) {
+			totalNegativeResultsSquared = totalNegativeResultsSquared.Add(movementPerCandle[x].Sub(riskFreeRatePerInterval).Pow(decimal.NewFromInt(2)))
+		}
+	}
+	if totalNegativeResultsSquared.IsZero() {
+		return decimal.Zero, ErrNoNegativeResults
+	}
+	f, exact := totalNegativeResultsSquared.Float64()
+	var err error
+	if !exact {
+		err = fmt.Errorf("%w from %v to %v", ErrInexactConversion, totalNegativeResultsSquared, f)
+	}
+	fAverageDownsideDeviation := math.Sqrt(f / float64(len(movementPerCandle)))
```

Comment:
Might be an idea to add in a len(movementPerCandle) == 0 check just to be on the safe side.